stablehlo/reference/Ops.cpp

```diff
@@ -429,50 +429,63 @@ Tensor evalSliceOp(const Tensor &operand, Index startIndices, Sizes strides,
+    // handles, and the comparator knows how to use these handles to fetch
+    // the actual input elements being compared.
+    Index inputsTogether(inputs[0].getShape()[adjustedDimension]);
+    std::iota(inputsTogether.begin(), inputsTogether.end(), 0);
+    auto comparatorTogether = [&](int64_t lhsHandle, int64_t rhsHandle) {
       SmallVector<Tensor> args;
-      auto lhsIndex = *offsetIt;
-      auto rhsIndex = *offsetIt;
-      lhsIndex[dimension] = lhs;
-      rhsIndex[dimension] = rhs;
-      for (const auto &indexedInput : llvm::enumerate(inputs)) {
-        auto type = comparator.front()
-                        .getArgumentTypes()[2 * indexedInput.index()]
-                        .dyn_cast<TensorType>();
-        auto lhs = Tensor(type);
-        auto rhs = Tensor(type);
-        lhs.set({}, indexedInput.value().get(lhsIndex));
-        rhs.set({}, indexedInput.value().get(rhsIndex));
-        args.push_back(lhs);
-        args.push_back(rhs);
+      auto lhsIndex = *resultIt;
+      auto rhsIndex = *resultIt;
+      lhsIndex[adjustedDimension] = lhsHandle;
+      rhsIndex[adjustedDimension] = rhsHandle;
+      for (const auto &input : inputs) {
+        auto argType = RankedTensorType::get({}, input.getElementType());
+        auto lhsEl = Tensor(argType);
+        auto rhsEl = Tensor(argType);
+        lhsEl.set({}, input.get(lhsIndex));
+        rhsEl.set({}, input.get(rhsIndex));
+        args.push_back(lhsEl);
+        args.push_back(rhsEl);
       }
       auto cmpResult = eval(comparator, args, &scope);
       return cmpResult[0].get({}).getBooleanValue();
     };
     if (isStable)
-      std::stable_sort(indices.begin(), indices.end(), compare);
+      std::stable_sort(inputsTogether.begin(), inputsTogether.end(),
+                       comparatorTogether);
     else
-      std::sort(indices.begin(), indices.end(), compare);
-    auto resultIdx = *offsetIt;
-    auto operandIdx = *offsetIt;
-    for (auto [dst, src] : llvm::enumerate(indices)) {
-      for (auto [input, result] : llvm::zip(inputs, runtimeResults)) {
-        operandIdx[dimension] = src;
-        resultIdx[dimension] = static_cast<int64_t>(dst);
-        result.set(resultIdx, input.get(operandIdx));
+      std::sort(inputsTogether.begin(), inputsTogether.end(),
+                comparatorTogether);
+
+    // After the vector of handles has been sorted, we apply the results of
+    // this sort by reshuffling input elements into result elements.
+    for (auto [inputHandle, resultHandle] : llvm::enumerate(inputsTogether)) {
+      for (auto [input, result] : llvm::zip(inputs, results)) {
+        auto inputIdx = *resultIt;
+        auto resultIdx = *resultIt;
+        inputIdx[adjustedDimension] = inputHandle;
+        resultIdx[adjustedDimension] = resultHandle;
+        result.set(resultIdx, input.get(inputIdx));
       }
     }
   }
-  return runtimeResults;
+  return results;
```

Comment:
I don't think that the "runtime" prefix buys us much (or anything at all, really).